pkg/chunk/storage/caching_index_client.go

```diff
@@ -134,13 +134,13 @@ func (s *cachingIndexClient) QueryPages(ctx context.Context, queries []chunk.Ind
 
 		results[key] = rb
 	}
-
+	var iter chunk.ReadBatchIterator
 	err = s.IndexClient.QueryPages(ctx, cacheableMissed, func(cacheableQuery chunk.IndexQuery, r chunk.ReadBatch) bool {
 		resultsMtx.Lock()
 		defer resultsMtx.Unlock()
 		key := queryKey(cacheableQuery)
 		existing := results[key]
-		for iter := r.Iterator(); iter.Next(); {
+		for iter = r.Iterator(iter); iter.Next(); {
```

Comment:
Bug: Reuses iterator between concurrent goroutines.

Reply:
I don't think that's true, but I agree that it's not super clear and may be dangerous to maintain. There's a lock.

Reply:
You are right, I missed the lock here.